lib/proxy/peer/quic.go

```diff
@@ -0,0 +1,141 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package peer
+
+import (
+	"time"
+
+	"github.com/quic-go/quic-go/quicvarint"
+)
+
+const (
+	quicMaxIdleTimeout  = 30 * time.Second
+	quicKeepAlivePeriod = 5 * time.Second
+
+	quicMaxReceiveWindow = quicvarint.Max
+
+	// quicNextProto is the ALPN indicator for the current version of the QUIC
+	// proxy peering protocol.
+	quicNextProto = "teleport-peer-v1a"
+
+	// quicMaxMessageSize is the maximum accepted size (in protobuf binary
+	// format) for the request and response messages exchanged as part of the
+	// dialing.
+	quicMaxMessageSize = 128 * 1024
```

Comment:
This is much smaller than the uint32 we keep mentioning. Why is that?

Reply:
Seems like a reasonable size, and if we tuned it down to 65535 and used a uint16 we wouldn't have any way to make it bigger without changing the protocol, if it came down to it.

Reply:
Got it. Let's mention that this is somewhat arbitrary and that the reason for the uint32 is that we want room to grow.

Should we limit the messages sent in the other direction to that size too?

Reply:
Both directions are limited now, we log a message with the ever helpful "(this is a bug)" instead of sending an oversized message.

Reply:
If we hit the bug and abort the send, what happens to the connection? Should we make an effort to truncate?